mailpoet/assets/js/src/segments/dynamic/subscribers_counter.tsx

```diff
@@ -24,34 +24,62 @@ const SubscribersCounter: React.FunctionComponent = () => {
   const { updateSubscriberCount } = useDispatch('mailpoet-dynamic-segments-form');
 
   const serializedSegment = JSON.stringify(segment);
+  const latestRequestIdRef = useRef(1);
+  const deferredRequestRef = useRef(null);
+  const isRequestInFlight = useRef(false);
+
+  function load(loadItem: Segment): void {
+    // Don't allow multiple in-flight requests to avoid hammering the database
+    // when we'll only ever use the results of the last request
+    if (isRequestInFlight.current) {
+      deferredRequestRef.current = loadItem;
+      return;
+    }
+
+    deferredRequestRef.current = null;
+    latestRequestIdRef.current += 1;
+    const requestId = latestRequestIdRef.current;
+    isRequestInFlight.current = true;
+
+    loadCount(loadItem).then((response) => {
+      isRequestInFlight.current = false;
+      if (deferredRequestRef.current) {
+        load(deferredRequestRef.current);
+        return;
+      }
```

Comment:
I'm wondering whether the deferring logic is needed. It seems to me `debounce` should handle that already, shouldn't it? Or is it because the requests are very slow? It adds quite a bit of complexity but I guess there's not much we can do as we don't have a better logic in the codebase for this (like request cancellation).

Reply:
@JanJakes I went back and forth on this myself. My [original idea](https://mailpoet.atlassian.net/browse/MAILPOET-4018?focusedCommentId=21914) was that we could use deferring instead of debouncing. That way fast-to-calculate counts would come back right away, but with long-running queries there could only ever be one request at a time. We also wouldn't need to keep track of the order of requests.

@costasovo [thought](https://mailpoet.atlassian.net/browse/MAILPOET-4018?focusedCommentId=21915) it could be best to combine these two approaches, but I do think having both adds more complexity than its worth.

Combined, deferring only protects against what is probably an edge case where someone is making many changes to a dynamic segment that takes a very long time to calculate. Our timeout for these requests is 20 seconds, so worst case scenario, if you were very carefully pausing every 2 seconds or so, you could have around 10 simultaneous Ajax requests.

If the query itself is extremely inefficient and takes significantly more than 20 seconds to execute then this could be a slightly bigger deal. 

So I think we should go with just one or the other, and here are my thoughts I jotted down while considering both:

## Deferred Only

### Pros
- No dependency on third party library for `debounce`
- No need to track order of requests because they will only ever be made sequentially
- Users potentially see feedback much faster
- Less potential for initiating many simultaneous, long-running queries
- Potentially fewer total queries when working on segments with very long running count calculations

### Cons
- Will always execute at least 2 queries when user makes many changes in quick succession
- Could result in a much higher number of total queries if calculations are fast
- Could take longer to get results when going from a very complex segment to a very simple one (the complex query that would start immediately would need to complete before the simple one could start.

## Debounce Only

### Pros
- Potentially much fewer total DB queries when the calculations are fast
- User might see results faster when changing from a complex segment to a very simple one

### Cons
- User always has to wait at least 2 seconds to see a new count
- Easier to generate simultaneous long-running queries

Deciding between these choices could require some additional information
1. How long are the long-running queries that have been causing issues?
2. Is total query count something we care about?
3. What's the average amount of time it takes to calculate these counts? More or less than 2 seconds?
4. Is it more common for existing segments to increase or decrease in complexity when users edit them?

I think I'm overthinking this. 

I tried both ways and pushed the branches. Here are comparisons with this PR for reference.
[Deferred only](https://github.com/mailpoet/mailpoet/compare/master...update/MAILPOET-4018/deferred-only?expand=1)
[Debounce only](https://github.com/mailpoet/mailpoet/compare/master...update/MAILPOET-4018/deferred-only?expand=1)

I still think I prefer the defer-only approach, but I'd like go defer (🤣) to someone with more experience with this part of the app.

Reply:
Thanks for the detailed answer, I guess we can also keep both approaches, I see it seems to make sense as well.